### PR TITLE
My Jetpack: Update checkout flow to connect "after" checkout vs before.

### PR DESCRIPTION
**Pull request by elliottprogrammer**
This PR updates the My Jetpack product checkout flow to connect Jetpack "after" checkout is completed as opposed to before arriving to checkout.

This was originally discussed in P2: "wp-admin → checkout flow simplification": p1HpG7-onP-p2
**TLDR;-**
A Jetpack plugin user should be able to make a product purchase (initiated from My Jetpack, when disconnected) without having to manually connect the plugin first, but rather should be able to purchase the product first, then have the connection automatically initiated afterward.

See screencast video below to view the full "connect _**After**_ checkout" flow in action. ⤵️

## Proposed changes:

* Added a `getCalypsoOrigin()` helper in the connection package.
* Updated the My Jetpack product checkout flow (if site not connected/registered) to instead link to a connectAfterCheckout flow in Calypso as opposed to a connect "before" checkout flow.

### Other information:

- [X] Have you written new tests for your changes, if applicable?
- [X] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [X] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Does this pull request change what data or activity we track or use?

No

## Testing instructions:

- Do one of these:
    - Checkout this PR and spin up a local development build of Jetpack, OR
    - Spin up a Jurassic.ninja site with the Jetpack Beta plugin running this PR.
- For minimal testing, skip to the section, "Basic Testing this PR only" below, otherwise follow the instructions to test the full checkout flow.

#### Testing the full checkout flow:
- Jetpack plugin activated, Not connected:
    - Go to My Jetpack:  `/wp-admin/admin.php?page=my-jetpack`
    - On the VaultPress Backup Card, click the "Purchase" Button.
    - Click "Get Jetpack VaultPress Backup" button.
    - Verify you are taken to: https://wordpress.com/checkout/jetpack/jetpack_backup_t1_yearly with at least these query args: `connect_after_checkout=true`, `admin_url=[your site admin url]`,  `from_site_slug=[your site slug]` and with Jetpack VaultPress Backup (10GB) in the cart. 
    - On the checkout page, use A81N credits, and select "Assign a payment method later".
    - Click "Complete checkout"
    - Verify purchase is completed and you are redirected to the Connection/Authorization screen.
    - Click "Authorize".
    - Verify connection completes successfully and you are redirected to Licensing Thank you auto-activation screen (`/checkout/jetpack/thank-you/licensing-auto-activate/jetpack_backup_t1_yearly?...`).
    -  See screencast video below to see the full "connect _After_ checkout" flow (these instructions) in action. ⤵️

#### Bonus points: - Test using PayPal

- Follow the instructions above, but instead of using A81N credits, use a test PayPal account (See: PCYsg-IA-p2  )

#### Basic Testing this PR only.
- Jetpack plugin activated, Not connected:
    - Go to:  `/wp-admin/admin.php?page=my-jetpack`
    - On the VaultPress Backup Card, click the "Purchase Button"
    - Click "Get Jetpack VaultPress Backup" button.
    - Verify you are taken to: https://wordpress.com/checkout/jetpack/jetpack_backup_t1_yearly with at least these query args: `connect_after_checkout=true`, `admin_url=[your site admin url]`,  `from_site_slug=[your site slug]`
 - Connect Jetpack:
     - Go to:  `/wp-admin/admin.php?page=my-jetpack`
    - On the VaultPress Backup Card, click the "Purchase Button"
    - Click "Get Jetpack VaultPress Backup" button.
    - Verify you are taken to: https://wordpress.com/checkout/[your site slug] (some query args will also be present) with Jetpack VaultPress Backup (10GB) in the cart. 

### Screencast: Connect _after_ checkout flow

https://github.com/Automattic/jetpack/assets/11078128/da0e540b-b722-4de3-927d-0952cdc3cbd5

### CODE REVIEW

File: projects/js-packages/connection/hooks/use-product-checkout-workflow/index.jsx
Comment (CodeyGuyDylan) [inline projects/js-packages/connection/hooks/use-product-checkout-workflow/index.jsx]: Think there is a typo here, should be `useConnectAfterCheckoutFlow`

**Review by CodeyGuyDylan — CHANGES_REQUESTED** [review]
The testing instructions worked well. I am wondering though, would it be possible to automatically assign the license to the site in this case? My understanding of this is to allow users to purchase licenses without having to have Jetpack connected but I would think the user would still expect that the license they just bought to be auto-assigned to the site they were purchasing it for 🤔 

Let me know if I am misunderstanding something

**Comment by jboland88** [comment]
>  would it be possible to automatically assign the license to the site in this case? My understanding of this is to allow users to purchase licenses without having to have Jetpack connected but I would think the user would still expect that the license they just bought to be auto-assigned to the site they were purchasing it for 🤔

@CodeyGuyDylan - I think that's the final intent, it sounds like the license assignment may be handled in a separate PR based on the PR description though:

> TODO: Then after connection complete, activate the product license key (transfer the temporary site subscription to the target site) - In a separate, follow-up diff/PR.

**Comment by CodeyGuyDylan** [comment]
> > would it be possible to automatically assign the license to the site in this case? My understanding of this is to allow users to purchase licenses without having to have Jetpack connected but I would think the user would still expect that the license they just bought to be auto-assigned to the site they were purchasing it for 🤔
> 
> @CodeyGuyDylan - I think that's the final intent, it sounds like the license assignment may be handled in a separate PR based on the PR description though:
> 
> > TODO: Then after connection complete, activate the product license key (transfer the temporary site subscription to the target site) - In a separate, follow-up diff/PR.

Oh shoot! I missed that, my apologies

File: projects/js-packages/connection/hooks/use-product-checkout-workflow/index.jsx
Comment (jboland88) [inline projects/js-packages/connection/hooks/use-product-checkout-workflow/index.jsx:77]: @sirbrillig mentioned this in the corresponding Calypso PR - I wonder if we can find a more descriptive name for this parameter 🤔 `site_to_connect` maybe?

**Review by CodeyGuyDylan — DISMISSED** [review]
After realizing the product activation would come separately in a later PR, I think this one is good to go other than one typo I found in a variable name

**Review by jboland88 — DISMISSED** [review]
This LGTM and tested well! I only had one naming nitpick.

File: projects/js-packages/connection/hooks/use-product-checkout-workflow/index.jsx
Comment (elliottprogrammer) [reply projects/js-packages/connection/hooks/use-product-checkout-workflow/index.jsx:77]: Yes, I agree `from_site_slug` is a bit confusing what it is...  Let me explain:
 `fromSiteSlug` is the Jetpack "site-slug passed _**from the site**_ via url query arg" (into Calypso), for use cases when we want to know the site but there is not a site in context, such as in siteless checkout. As opposed to Calypso's use of `siteSlug` which is the site-slug present when the site is in context (ie- the site is available in State, such as when the site is connected and the user is logged in).
My only concern with changing the variable name to `site_to_connect` is that it restricts it's use case to only connecting the site, when the variable could also potentially be used for other purposes in which we need to know the site but we cannot get the selected site from State for whatever reason (when the site not connected, user not logged-in, etc).
Addressing @sirbrilling's feedback in the corresponding Calypso PR, I added comments in all the places the variable is present in Calypso, clearly describing what the variable is and it's purpose, so for now I will do the same thing here- I'll add a comment, clearly describing what the variable/arg is and what it's used for.
That being said, if you still feel strongly about changing the variable to `site_to_connect`, I'm not completely opposed to it, so please let me know your thoughts and I can change it here and all throughout the Calypso PR if it's decided that would be best.

Comment (jboland88) [reply projects/js-packages/connection/hooks/use-product-checkout-workflow/index.jsx:77]: Aha! thanks for that extra context! That makes sense and `site_to_connect` does not seem to be the best fit.

I think adding comments should help clarify 👍

File: projects/js-packages/connection/hooks/use-product-checkout-workflow/index.jsx
Comment (IanRamosC) [inline projects/js-packages/connection/hooks/use-product-checkout-workflow/index.jsx]: This is probably a minor typo with a tab instead of space?

**Review by IanRamosC — DISMISSED** [review]
Code looks good and tests well. There's one small difference from your test plan, though. Right after clicking to approve the user connection, I'm redirected to the usual https://wordpress.com/jetpack/connect/plans/, instead of My Jetpack. Other than that, it works as described

image

**Comment by elliottprogrammer** [comment]
~[DO NOT MERGE] yet...~
~We don't wan't to merge this PR until https://github.com/Automattic/wp-calypso/pull/82751 is approved and shipped, and we've thoroughly tested this (connect-**After**-checkout) checkout flow.~

File: projects/js-packages/connection/hooks/use-product-checkout-workflow/index.jsx
Comment (elliottprogrammer) [reply projects/js-packages/connection/hooks/use-product-checkout-workflow/index.jsx]: Yes, fixed. 👍🙂

**Comment by elliottprogrammer** [comment]
@IanRamosC, This should be good to go now. I was wondering if you could give this a final look/review?

**Comment by sergeymitr** [comment]
Works like a charm!

There seems to be room for improvement on Calypso side:
Screenshot 2023-10-20 at 16 56 47

I feel like asking users to login is a bit excessive, since new users can finish checkout just fine.

**Review by sergeymitr — DISMISSED** [review]

**Comment by elliottprogrammer** [comment]
> Works like a charm!
> 
> There seems to be room for improvement on Calypso side: 
> I feel like asking users to login is a bit excessive, since new users can finish checkout just fine.

Thanks for reviewing @sergeymitr!  Yes agreed, we have a follow-up project to address this: 😉👍https://github.com/orgs/Automattic/projects/702/views/6?pane=issue&itemId=39392983

**Review by sergeymitr — APPROVED** [review]
Re-approving 👍

**Pull request merged**
